spec/contracts/base_contract_spec.rb

```diff
@@ -0,0 +1,261 @@
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2024 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+require "spec_helper"
+
+RSpec.describe BaseContract do
+  let(:model) { double(name: "name") } # rubocop:disable RSpec/VerifiedDoubles
+  let(:user) { build(:user) }
+
+  describe ".writable_attributes" do
+    it "returns attributes with `writable: true`" do
+      contract_class = Class.new(BaseContract) do
+        attribute :name, writable: true
+      end
+      expect(contract_class.writable_attributes).to contain_exactly("name", "name_id")
+    end
+
+    it "does not return attributes with `writable: false`" do
+      contract_class = Class.new(BaseContract) do
+        attribute :name, writable: false
+      end
+      expect(contract_class.writable_attributes).to be_empty
+    end
+
+    it "returns attributes with `writable: nil`" do
+      contract_class = Class.new(BaseContract) do
+        attribute :name, writable: nil
+      end
+      expect(contract_class.writable_attributes).to contain_exactly("name", "name_id")
+    end
+
+    it "returns attributes without `:writable` parameter (same as `writable: nil`)" do
+      contract_class = Class.new(BaseContract) do
+        attribute :name
+      end
+      expect(contract_class.writable_attributes).to contain_exactly("name", "name_id")
+    end
+
+    it "returns attributes with `writable: -> { true }`" do
+      contract_class = Class.new(BaseContract) do
+        attribute :name, writable: -> { true }
+      end
+      expect(contract_class.writable_attributes).to contain_exactly("name", "name_id")
+    end
+
+    it "returns attributes with `writable: -> { false }`" do
+      contract_class = Class.new(BaseContract) do
+        attribute :name, writable: -> { false }
+      end
+      expect(contract_class.writable_attributes).to contain_exactly("name", "name_id")
+    end
```

Comment:
Why?

Reply:
Because it's not being instantiated and therefore only prevents adding to `writable_attributes` if it's directly a primitive boolean? 

From the code, I see it only execs the lambdas when a contract is instantiated and now needs to look in the hierarchy and eval the lambda

Reply:
> Why?

I don't know, I only tested the existing behavior. I think evaluating the lambda only makes sense when you have a contract instance, and not a class. The class method `.writable_attributes` does not seem very useful IMHO.